File: app/src/main/java/code/name/monkey/retromusic/ui/activities/EqualizerActivity.java
package code.name.monkey.retromusic.ui.activities;

import android.os.Bundle;
import com.google.android.material.appbar.AppBarLayout;
import androidx.appcompat.widget.SwitchCompat;
import androidx.appcompat.widget.Toolbar;
import android.transition.TransitionManager;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.CompoundButton;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.Spinner;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.ButterKnife;
import code.name.monkey.appthemehelper.ThemeStore;
import code.name.monkey.retromusic.R;
import code.name.monkey.retromusic.helper.EqualizerHelper;
import code.name.monkey.retromusic.ui.activities.base.AbsMusicServiceActivity;

/**
 * @author Hemanth S (h4h13).
 */

public class EqualizerActivity extends AbsMusicServiceActivity implements
        AdapterView.OnItemSelectedListener {

    @BindView(R.id.equalizer)
    SwitchCompat mEnable;

    @BindView(R.id.content)
    LinearLayout mContent;

    @BindView(R.id.bands)
    LinearLayout mLinearLayout;

    @BindView(R.id.bass_boost_strength)
    SeekBar mBassBoostStrength;

    @BindView(R.id.virtualizer_strength)
    SeekBar mVirtualizerStrength;

    @BindView(R.id.bass_boost)
    TextView mBassBoost;

    @BindView(R.id.virtualizer)
    TextView mVirtualizer;

    @BindView(R.id.toolbar)
    Toolbar mToolbar;

    @BindView(R.id.app_bar)
    AppBarLayout mAppBar;

    @BindView(R.id.presets)
    Spinner mPresets;

    @BindView(R.id.title)
    TextView mTitle;

    private CompoundButton.OnCheckedChangeListener mListener = (buttonView, isChecked) -> {
        switch (buttonView.getId()) {
            case R.id.equalizer:
                EqualizerHelper.getInstance().getEqualizer().setEnabled(isChecked);
                TransitionManager.beginDelayedTransition(mContent);
                mContent.setVisibility(isChecked ? View.VISIBLE : View.GONE);
                break;
        }
    };
    private SeekBar.OnSeekBarChangeListener mSeekBarChangeListener = new SeekBar.OnSeekBarChangeListener() {
        @Override
        public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
            if (fromUser) {
                if (seekBar == mBassBoostStrength) {
                    mBassBoost.setEnabled(progress > 0);
                    EqualizerHelper.getInstance().setBassBoostStrength(progress);
                    EqualizerHelper.getInstance().setBassBoostEnabled(progress > 0);
                } else if (seekBar == mVirtualizerStrength) {
                    mVirtualizer.setEnabled(progress > 0);
                    EqualizerHelper.getInstance().setVirtualizerEnabled(progress > 0);
                    EqualizerHelper.getInstance().setVirtualizerStrength(progress);
                }
            }
        }

        @Override
        public void onStartTrackingTouch(SeekBar seekBar) {

        }

        @Override
        public void onStopTrackingTouch(SeekBar seekBar) {

        }
    };
    private ArrayAdapter<String> mPresetsNamesAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_equalizer);
        ButterKnife.bind(this);

        setStatusbarColorAuto();
        setNavigationbarColorAuto();
        setTaskDescriptionColorAuto();
        setLightNavigationBar(true);

        setupToolbar();

        mEnable.setChecked(EqualizerHelper.getInstance().getEqualizer().getEnabled());
        mEnable.setOnCheckedChangeListener(mListener);

        mPresetsNamesAdapter = new ArrayAdapter<String>(this, android.R.layout.simple_list_item_1);
        mPresets.setAdapter(mPresetsNamesAdapter);
        mPresets.setOnItemSelectedListener(this);

        mBassBoostStrength.setProgress(EqualizerHelper.getInstance().getBassBoostStrength());
        mBassBoostStrength.setOnSeekBarChangeListener(mSeekBarChangeListener);

        mVirtualizerStrength.setProgress(EqualizerHelper.getInstance().getVirtualizerStrength());
        mVirtualizerStrength.setOnSeekBarChangeListener(mSeekBarChangeListener);

        setupUI();
        addPresets();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void setupToolbar() {
        mTitle.setTextColor(ThemeStore.textColorPrimary(this));
        int primaryColor = ThemeStore.primaryColor(this);
        mToolbar.setBackgroundColor(primaryColor);
        mAppBar.setBackgroundColor(primaryColor);
        mToolbar.setNavigationIcon(R.drawable.ic_keyboard_backspace_black_24dp);
        mToolbar.setNavigationOnClickListener(v -> onBackPressed());
        setSupportActionBar(mToolbar);
        setTitle(null);
    }

    private void addPresets() {
        mPresetsNamesAdapter.clear();
        mPresetsNamesAdapter.add("Custom");
        for (int j = 0; j < EqualizerHelper.getInstance().getEqualizer().getNumberOfPresets(); j++) {
            mPresetsNamesAdapter
                    .add(EqualizerHelper.getInstance().getEqualizer().getPresetName((short) j));
            mPresetsNamesAdapter.notifyDataSetChanged();
        }
        mPresets
                .setSelection((int) EqualizerHelper.getInstance().getEqualizer().getCurrentPreset() + 1);
    }

    private void setupUI() {
        mLinearLayout.removeAllViews();
        short bands;
        try {
            // get number of supported bands
            bands = (short) EqualizerHelper.getInstance().getNumberOfBands();

            // for each of the supported bands, we will set up a slider from -10dB to 10dB boost/attenuation,
            // as well as text labels to assist the user
            for (short i = 0; i < bands; i++) {
                final short band = i;

                View view = LayoutInflater.from(this).inflate(R.layout.retro_seekbar, mLinearLayout, false);
                TextView freqTextView = view.findViewById(R.id.hurtz);
                freqTextView.setText(
                        String.format("%d Hz", EqualizerHelper.getInstance().getCenterFreq((int) band) / 1000));

                TextView minDbTextView = view.findViewById(R.id.minus_db);
                minDbTextView
                        .setText(String.format("%d dB", EqualizerHelper.getInstance().getBandLevelLow() / 100));

                TextView maxDbTextView = view.findViewById(R.id.plus_db);
                maxDbTextView.setText(
                        String.format("%d dB", EqualizerHelper.getInstance().getBandLevelHigh() / 100));

                SeekBar bar = view.findViewById(R.id.seekbar);
                bar.setMax(EqualizerHelper.getInstance().getBandLevelHigh() - EqualizerHelper.getInstance()
                        .getBandLevelLow());
                bar.setProgress(
                        EqualizerHelper.getInstance().getBandLevel((int) band) - EqualizerHelper.getInstance()
                                .getBandLevelLow());
                bar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
                    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                        EqualizerHelper.getInstance().setBandLevel((int) band,
                                (int) (progress + EqualizerHelper.getInstance().getBandLevelLow()));
                        if (fromUser) {
                            mPresets.setSelection(0);
                        }
                    }

                    public void onStartTrackingTouch(SeekBar seekBar) {
                    }

                    public void onStopTrackingTouch(SeekBar seekBar) {
                    }
                });

                mLinearLayout.addView(view);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        if (position == 0) {
            return;
        }
        EqualizerHelper.getInstance().getEqualizer().usePreset((short) (position - 1));
        setupUI();
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {

    }
}

File: app/src/main/java/code/name/monkey/retromusic/ui/activities/EqualizerActivity.kt
package code.name.monkey.retromusic.ui.activities

import android.os.Bundle
import android.transition.TransitionManager
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.SeekBar
import android.widget.TextView
import code.name.monkey.appthemehelper.ThemeStore
import code.name.monkey.appthemehelper.util.TintHelper
import code.name.monkey.appthemehelper.util.ToolbarContentTintHelper
import code.name.monkey.retromusic.R
import code.name.monkey.retromusic.helper.EqualizerHelper
import code.name.monkey.retromusic.ui.activities.base.AbsMusicServiceActivity
import code.name.monkey.retromusic.util.ViewUtil
import kotlinx.android.synthetic.main.activity_equalizer.*

/**
 * @author Hemanth S (h4h13).
 */

class EqualizerActivity : AbsMusicServiceActivity(), AdapterView.OnItemSelectedListener {


    /*private val mListener = { buttonView, isChecked ->
        when (buttonView.getId()) {
            R.id.equalizerSwitch -> {
                EqualizerHelper.instance!!.equalizer.enabled = isChecked
                TransitionManager.beginDelayedTransition(content)
                content.visibility = if (isChecked) View.VISIBLE else View.GONE
            }
        }
    }*/
    private val mSeekBarChangeListener = object : SeekBar.OnSeekBarChangeListener {
        override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
            if (fromUser) {
                if (seekBar === bassBoostStrength) {
                    bassBoost.isEnabled = progress > 0
                    EqualizerHelper.instance!!.bassBoostStrength = progress
                    EqualizerHelper.instance!!.isBassBoostEnabled = progress > 0
                } else if (seekBar === virtualizerStrength) {
                    virtualizer.isEnabled = progress > 0
                    EqualizerHelper.instance!!.isVirtualizerEnabled = progress > 0
                    EqualizerHelper.instance!!.virtualizerStrength = progress
                }
            }
        }

        override fun onStartTrackingTouch(seekBar: SeekBar) {

        }

        override fun onStopTrackingTouch(seekBar: SeekBar) {

        }
    }


    private var mPresetsNamesAdapter: ArrayAdapter<String>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_equalizer)


        setStatusbarColorAuto()
        setNavigationbarColorAuto()
        setTaskDescriptionColorAuto()
        setLightNavigationBar(true)

        setupToolbar()

        equalizerSwitch.isChecked = EqualizerHelper.instance!!.equalizer.enabled
        TintHelper.setTintAuto(equalizerSwitch, ThemeStore.accentColor(this), false)
        equalizerSwitch.setOnCheckedChangeListener { buttonView, isChecked ->
            when (buttonView.id) {
                R.id.equalizerSwitch -> {
                    EqualizerHelper.instance!!.equalizer.enabled = isChecked
                    TransitionManager.beginDelayedTransition(content)
                    content.visibility = if (isChecked) View.VISIBLE else View.GONE
                }
            }
        }

        mPresetsNamesAdapter = ArrayAdapter(this, android.R.layout.simple_list_item_1)
        presets.adapter = mPresetsNamesAdapter
        presets.onItemSelectedListener = this

        bassBoostStrength.progress = EqualizerHelper.instance!!.bassBoostStrength
        ViewUtil.setProgressDrawable(bassBoostStrength, ThemeStore.accentColor(this))
        bassBoostStrength.setOnSeekBarChangeListener(mSeekBarChangeListener)

        virtualizerStrength.progress = EqualizerHelper.instance!!.virtualizerStrength
        ViewUtil.setProgressDrawable(virtualizerStrength, ThemeStore.accentColor(this))
        virtualizerStrength.setOnSeekBarChangeListener(mSeekBarChangeListener)

        setupUI()
        addPresets()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressed()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun setupToolbar() {
        bannerTitle.setTextColor(ThemeStore.textColorPrimary(this))
        val primaryColor = ThemeStore.primaryColor(this)
        appBarLayout.setBackgroundColor(primaryColor)
        toolbar.apply {
            setBackgroundColor(primaryColor)
            setNavigationIcon(R.drawable.ic_keyboard_backspace_black_24dp)
            setNavigationOnClickListener { onBackPressed() }
            setSupportActionBar(this)
            ToolbarContentTintHelper.colorBackButton(this, ThemeStore.textColorSecondary(context))
        }
        title = null
    }

    private fun addPresets() {
        mPresetsNamesAdapter!!.clear()
        mPresetsNamesAdapter!!.add("Custom")
        for (j in 0 until EqualizerHelper.instance!!.equalizer.numberOfPresets) {
            mPresetsNamesAdapter!!
                    .add(EqualizerHelper.instance!!.equalizer.getPresetName(j.toShort()))
            mPresetsNamesAdapter!!.notifyDataSetChanged()
        }
        presets.setSelection(EqualizerHelper.instance!!.equalizer.currentPreset.toInt() + 1)
    }

    private fun setupUI() {
        frequencyBands.removeAllViews()
        val bands: Short
        try {
            // get number of supported bands
            bands = EqualizerHelper.instance!!.numberOfBands.toShort()

            // for each of the supported bands, we will set up a slider from -10dB to 10dB boost/attenuation,
            // as well as text labels to assist the user
            for (i in 0 until bands) {

                val view = LayoutInflater.from(this).inflate(R.layout.retro_seekbar, frequencyBands, false)
                val freqTextView = view.findViewById<TextView>(R.id.hurtz)
                freqTextView.text = String.format("%d Hz", EqualizerHelper.instance!!.getCenterFreq(i) / 1000)

                val minDbTextView = view.findViewById<TextView>(R.id.minus_db)
                minDbTextView.text = String.format("%d dB", EqualizerHelper.instance!!.bandLevelLow / 100)

                val maxDbTextView = view.findViewById<TextView>(R.id.plus_db)
                maxDbTextView.text = String.format("%d dB", EqualizerHelper.instance!!.bandLevelHigh / 100)

                val bar = view.findViewById<SeekBar>(R.id.seekbar)
                ViewUtil.setProgressDrawable(bar, ThemeStore.accentColor(this))
                bar.max = EqualizerHelper.instance!!.bandLevelHigh - EqualizerHelper.instance!!
                        .bandLevelLow
                bar.progress = EqualizerHelper.instance!!.getBandLevel(i) - EqualizerHelper.instance!!
                        .bandLevelLow
                bar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
                    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                        EqualizerHelper.instance!!.setBandLevel(i,
                                progress + EqualizerHelper.instance!!.bandLevelLow)
                        if (fromUser) {
                            presets.setSelection(0)
                        }
                    }

                    override fun onStartTrackingTouch(seekBar: SeekBar) {}

                    override fun onStopTrackingTouch(seekBar: SeekBar) {}
                })

                frequencyBands.addView(view)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        if (position == 0) {
            return
        }
        EqualizerHelper.instance!!.equalizer.usePreset((position - 1).toShort())
        setupUI()
    }

    override fun onNothingSelected(parent: AdapterView<*>) {

    }
}
